### PR TITLE
fix: hide dropzone when drop happens outside

### PR DESCRIPTION
**Pull request by giuliano176**
Refs: CDS-193

### CODE REVIEW

File: src/components/utilities/Drop.tsx
Comment (beawar) [inline src/components/utilities/Drop.tsx]: By removing the generic, you are introducing an implicit any for the event argument. Please add the right type

File: src/components/utilities/Drop.tsx
Comment (beawar) [inline src/components/utilities/Drop.tsx]: Since the function returns a boolean, I feel the `get` is not necessary in the name, and also misleading. A `targetIsOverlayElement`, or `isTargetOverlayElement` seems clearer to me.

File: src/components/utilities/Drop.tsx
Comment (beawar) [inline src/components/utilities/Drop.tsx]: Instead of casting to HTMLElement, use instanceof, which allows you to check for the real type at runtime. The cast is forcing the type in a wrong way, it is not considering the undefined type, even if you are checking it at line 57. Remove the cast and update line 57 to use the instanceof check

File: src/components/utilities/Drop.tsx
Comment (beawar) [inline src/components/utilities/Drop.tsx]: `e` is an implicit any. Add the right type

**Review by beawar — COMMENTED** [review]
The dropzone is flickering while moving the mouse over it during the drag. You can see in the video the difference from before, where this wasn't happening. The flickering problem is also present in mails, even if it is not so evident, I think because there are additional layers which are increasing the complexity of the component (second video).

https://github.com/zextras/carbonio-design-system/assets/17934172/7be21b77-ad77-48e4-91f5-7b22e32871c3

https://github.com/zextras/carbonio-design-system/assets/17934172/84cbae9b-1617-4e21-b8fe-ffda2eafbe32


I feel like the core problem of the Drop component is that it is not considering the possibility that a drop can happen outside the dropzone itself. The Drop is not removing the dropzone entirely when a drag leave happens, it is just removing the graphically visible part. It has probably more sense to use a state to decide whether the entire dropzone should be visible or not, instead of relying on that window.draggedItem.

Also, be careful on removing the the pointer-events set to none: the goal of that rule is to prevent the children to capture the events during the drag, if not explicitly forced with an additional pointer-events rule set on the child itself.

File: src/components/utilities/Drop.tsx
Comment (gnekoz) [inline src/components/utilities/Drop.tsx]: I think that, in this context, the usage of the _includes_ method is not 100% safe nor efficient.
The outerHTML contains the tag itself and all of its content. If the target isn't a _overlay-element_ but it contains it (or it contains the "overlay-element" string) the _isOverlayElement_ return true anyway.
Moreover if the target contains a large content (let's say a _img_ with a base64 content source) the _includes_ comparison will take some time.
Probably we can better achieve the same result by checking the _target.tagName_ property

Comment (giuliano176) [reply src/components/utilities/Drop.tsx]: I did struggle to find a solution to detect the element, and I agree with you that an html property read on the outerHTML attribute is not fail proof. but in absence of alternatives I think it's reasonable enough for our goal, tagName is not viable as it will always return DIV, so it won't help us.

in terms of safety, the corner cases that you are describing are in my view very unlikely to occur, I can however change the OVERLAY_ELEMENT_IDENTIFIER value to a more complex string?

i get your point in the search speed, I can maybe replace the includes with a regex test, which could have a somehow better performance?

Comment (gnekoz) [reply src/components/utilities/Drop.tsx]: Whoops, I missed the fact that searched string is the value of an attribute and not the element tag name, sorry about that.
So, instead of the tag name, the specific attribute should be inspected. This can be achieved by using the _attributes_ property of the element.
Something like:
`target.attributes.getNamedItem('type') `
should return the value of the attribute that we need to check

Comment (giuliano176) [reply src/components/utilities/Drop.tsx]: that is returning null, the value is not being read. the attributes property does not have a type attribute

File: src/components/utilities/Drop.tsx
Comment (beawar) [inline src/components/utilities/Drop.tsx]: ContainerProps type is already inherited from the Container itself. Re-declaring it is redundant.

I don't like the fact that the type attribute is set to distinguish between elements. I would prefer a data-type instead, which tells more about the fact that that value is a custom attribute.

**Review by beawar — COMMENTED** [review]
I don't understand why we need to early return inside the drag events if the event target is the overlay. Moreover through an hack which is not bulletproof. It is really not clear to me what is the purpose of those checks.

The problem was that the overlay wasn't removed if a drop happened outside the dropzone. And that overlay was preventing every other event on the children of the Drop.
Now with the dragging state, the overlay is removed on dragLeave, and so the overlay is removed not only on a drop inside the dropzone, but also when the drag exit it. In this scenario, I don't follow why we need to check who fired the drag event. Especially here where the dragover event, which is the most fired, always update the dragging state. Am I not seeing something? A specific wrong behavior? If so, could you explain it to me, and also what I need to do to reprocude it (if possible, create a test on it)?

**Comment by giuliano176** [comment]
> I don't understand why we need to early return inside the drag events if the event target is the overlay. Moreover through an hack which is not bulletproof. It is really not clear to me what is the purpose of those checks.
> 
> The problem was that the overlay wasn't removed if a drop happened outside the dropzone. And that overlay was preventing every other event on the children of the Drop. Now with the dragging state, the overlay is removed on dragLeave, and so the overlay is removed not only on a drop inside the dropzone, but also when the drag exit it. In this scenario, I don't follow why we need to check who fired the drag event. Especially here where the dragover event, which is the most fired, always update the dragging state. Am I not seeing something? A specific wrong behavior? If so, could you explain it to me, and also what I need to do to reprocude it (if possible, create a test on it)?

The problem with this bug isn't actually related to the dragging state. The check for which component is triggering the drag event is essential to prevent a loop of dragEnter and dragLeave. Once the overlay is applied by the dragEnter, a dragLeave event is triggered by the Children component, causing the removal of the overlay, and the cycle begins anew.
To reproduce it, you can simply remove the function "isOverlayElement" and its implementations and observe the loop occurring.

I agree with you that it's a workaround. I spent a good amount of time trying to find a solution, and this is the best I could come up with. It might not be the prettiest, but it gets the job done. If you have any other ideas or a different angle to sort out this issue, I'd be really happy to hear them. I am going with this solution mainly for time constraints as the fix is holding up a much awaited feature

**Comment by beawar** [comment]
> The check for which component is triggering the drag event is essential to prevent a loop of dragEnter and dragLeave. Once the overlay is applied by the dragEnter, a dragLeave event is triggered by the Children component, causing the removal of the overlay, and the cycle begins anew. To reproduce it, you can simply remove the function "isOverlayElement" and its implementations and observe the loop occurring.

Ok I see now it only affects chrome, I was trying in firefox and it wasn't giving the same loop problem :confounded:

I'm thinking if there is a different solution, I'll tell you asap if something comes to my mind (better if with a pr pointing to yours)

**Review by beawar — COMMENTED** [review]
The solution to avoid the loop is to check the relatedTarget of the event, which contains the element exited from. See the documentation of the MouseEvent https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/relatedTarget

With the relatedTarget, we can check in the dragLeave if the drag is leaving the dropzone, and hiding the overlay only in that case.

I'm going to create a PR

**Review by beawar — DISMISSED** [review]
I want to wait for other 2 reviews other then mine before merging this PR, since I contributed directly

File: src/components/utilities/Drop.md
Comment (CataldoMazzilli) [inline src/components/utilities/Drop.md]: I suggest to improve documented cases to show props usage

**Review by CataldoMazzilli — DISMISSED** [review]

**Review by CataldoMazzilli — APPROVED** [review]

**Review by rodleyorosa — APPROVED** [review]

**Pull request merged**
